### ISSUE TITLE
Generators are not supported by TokenStream

### ISSUE THREAD

**Issue by QuentinPerez** [rust-lang/rust#123191]
I tried this code:

```rust
#[proc_macro_attribute]
pub fn test_generatror(
    _: proc_macro::TokenStream,
    _: proc_macro::TokenStream,
) -> proc_macro::TokenStream {
    quote! {
        pub async gen fn my_gen_fn() -> u32 {
            0
        }
    }
    .into()
}
```

I expected to see this happen: No error

Instead, this happened: I got this error

```rust
#[test_generatror]
   | ^^^^^^^^^^^^ the visibility
   |
   = help: you likely meant to define an item, e.g., `pub fn foo() {}`
   = note: this error originates in the attribute macro `test_generatror` (in Nightly builds, run with -Z macro-backtrace for more info)
```

If bandwidth is limited, I'm willing to attempt integration. Could you please provide some guidance on how to proceed?

### Meta

`rustc --version --verbose`:
```
rustc 1.79.0-nightly (1388d7a06 2024-03-20)
binary: rustc
commit-hash: 1388d7a069d872bcfe5e5dd97ef61fa0a586fac0
commit-date: 2024-03-20
host: aarch64-apple-darwin
release: 1.79.0-nightly
LLVM version: 18.1.2
```

**Comment by Nilstrieb** [comment]
That's not quite the entire error, could you post that, including a full reproduction? I don't think there is anything about proc macros that is wrong here... are you using `syn`?

**Comment by taiki-e** [comment]
`pub async gen fn` is not supported in the first place, right? The error can be reproduced without TokenStream: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=9680400c367f31fd92047c1b422bee9a

**Comment by fmease** [comment]
You need to use Rust 2024.

**Comment by taiki-e** [comment]
Ah, I missed that. In that case, I guess the problem may be that the crate that defining macro is pre-2024 edition since call_site span inherits the edition of the caller. (Although, I do not know what the edition actually is, as a complete reproduction is not provided.)

**Comment by QuentinPerez** [comment]
> You need to use Rust 2024.

You right, I forgot to add it in the proc macro lib. Thank you !
